File: js/laberinto4_Inteligente.js
/**
 * LABERINTO4_INTELIGENTE.JS
 * Traducción de laberinto4_Inteligente.py a JavaScript.
 * LOS LABERINTOS COMO PUERTA DE ENTRADA A LA INTELIGENCIA ARTIFICIAL
 *
 * Muestra de forma animada cómo funcionan DFS y BFS para resolver laberintos,
 * evidenciando las ventajas y desventajas de cada uno.
 *
 * Uso desde el HTML:
 *   import { iniciarAnimacion } from './laberinto4_Inteligente.js';
 *   await iniciarAnimacion(canvas);
 */

import {
    Config,
    compartirGlobales,
    activarMultisalida,
    desactivarAnimacion,
    generarLaberinto,
    dibujarLaberinto,
    borrarParedes,
    sleep,
    ARRIBA, ABAJO, IZQUIERDA, DERECHA,
} from './laberinto0.js';


// ─────────────────────────────────────────────
// CONSTANTES
// ─────────────────────────────────────────────

const FPS = 24;

const BORDE        = 4;
const DOBLEBORDE   = BORDE * 2;
const ANCHO_LIENZO = 1000;
const ALTO_LIENZO  = 600;
const ANCHO_PANTALLA = ANCHO_LIENZO + DOBLEBORDE;
const ALTO_PANTALLA  = ALTO_LIENZO  + DOBLEBORDE;

const TAM_LADOS    = [100, 100, 50, 50, 50];
const CANT_NIVELES = TAM_LADOS.length;

const COLOR_FONDO    = '#141414';       // (20, 20, 20)
const COLOR_PARED    = '#32c832';       // (50, 200, 50)
const COLOR_JUGADOR  = '#f0c864';       // (240, 200, 100)
const COLOR_LLEGADA  = '#fafafa';       // (250, 250, 250)
const COLOR_SOLUCION = '#faa01e';       // (250, 160, 30)
const ALPHA_FONDO    = 'rgba(20,20,20,0.47)';  // equivalente a (20,20,20,120) sobre 255


// ─────────────────────────────────────────────
// ESTADO INTERNO
// ─────────────────────────────────────────────

let LADO, N_FILAS, N_COLUMNAS, JUGADOR_RADIO, JUGADOR_DIAMETRO;


// ─────────────────────────────────────────────
// API PÚBLICA
// ─────────────────────────────────────────────

export async function iniciarAnimacion(canvas) {

    canvas.width  = ANCHO_PANTALLA;
    canvas.height = ALTO_PANTALLA;
    desactivarAnimacion();
    activarMultisalida();

    for (let nivel = 0; nivel < CANT_NIVELES; nivel++) {

        // Configurar dimensiones para este nivel
        LADO           = TAM_LADOS[nivel];
        N_FILAS        = Math.floor(ALTO_LIENZO / LADO);
        N_COLUMNAS     = Math.floor(ANCHO_LIENZO / LADO);
        JUGADOR_RADIO  = Math.floor(LADO / 4);
        JUGADOR_DIAMETRO = JUGADOR_RADIO * 2;

        compartirGlobales(canvas, N_FILAS, N_COLUMNAS, BORDE, LADO, COLOR_FONDO, COLOR_PARED);

        const lab = await generarLaberinto(nivel);

        // Punto final aleatorio (alejado del origen para que la búsqueda sea interesante)
        const finalFil = Math.floor(Math.random() * (N_FILAS - 3)) + 3;
        const finalCol = Math.floor(Math.random() * (N_COLUMNAS - 3)) + 3;

        let DFStotal = 0, DFSsolucion = 0, BFStotal = 0, BFSsolucion = 0;

        // Ejecutar DFS (tipo 0) y luego BFS (tipo 1)
        for (let tipoDeBusqueda = 0; tipoDeBusqueda < 2; tipoDeBusqueda++) {

            let { posX, posY, posFil, posCol, finalX, finalY, distancia, porAnalizar, yaConocido }
                = await armarEscenario(canvas, lab, finalFil, finalCol);

            // Búsqueda principal
            while (!(posFil === finalFil && posCol === finalCol)) {

                porAnalizar.delete(`${posFil},${posCol}`);
                yaConocido.set(`${posFil},${posCol}`, distancia);

                if (posFil > 0 &&
                    !lab[posFil][posCol].paredes.includes(ARRIBA) &&
                    !lab[posFil-1][posCol].paredes.includes(ABAJO) &&
                    !yaConocido.has(`${posFil-1},${posCol}`))
                    porAnalizar.set(`${posFil-1},${posCol}`, distancia + 1);

                if (posFil < N_FILAS - 1 &&
                    !lab[posFil][posCol].paredes.includes(ABAJO) &&
                    !lab[posFil+1][posCol].paredes.includes(ARRIBA) &&
                    !yaConocido.has(`${posFil+1},${posCol}`))
                    porAnalizar.set(`${posFil+1},${posCol}`, distancia + 1);

                if (posCol > 0 &&
                    !lab[posFil][posCol].paredes.includes(IZQUIERDA) &&
                    !lab[posFil][posCol-1].paredes.includes(DERECHA) &&
                    !yaConocido.has(`${posFil},${posCol-1}`))
                    porAnalizar.set(`${posFil},${posCol-1}`, distancia + 1);

                if (posCol < N_COLUMNAS - 1 &&
                    !lab[posFil][posCol].paredes.includes(DERECHA) &&
                    !lab[posFil][posCol+1].paredes.includes(IZQUIERDA) &&
                    !yaConocido.has(`${posFil},${posCol+1}`))
                    porAnalizar.set(`${posFil},${posCol+1}`, distancia + 1);

                // DFS: toma el último agregado (stack)
                // BFS: toma el primero (queue)
                // Map preserva el orden de inserción, igual que dict en Python 3.7+
                let proximo;
                if (tipoDeBusqueda === 0) {
                    const entries = [...porAnalizar.entries()];
                    proximo = entries[entries.length - 1];  // último → DFS
                } else {
                    proximo = porAnalizar.entries().next().value;  // primero → BFS
                }

                const [clave, dist] = proximo;
                const [nFil, nCol] = clave.split(',').map(Number);
                distancia = dist;
                posFil = nFil;
                posCol = nCol;

                ({ posX, posY } = await moverBuscador(canvas, posX, posY, posFil, posCol, COLOR_JUGADOR));
            }


            yaConocido.set(`${posFil},${posCol}`, distancia);
            const total = yaConocido.size - 1;
            const camino = procesarRecorrido(yaConocido, lab);
            await caminarSolucion(canvas, camino);

            if (tipoDeBusqueda === 0) {
                DFStotal    = total;
                DFSsolucion = camino.length;
            } else {
                BFStotal    = total;
                BFSsolucion = camino.length;
            }
        }

        await presentarResultados(canvas, lab, DFStotal, DFSsolucion, BFStotal, BFSsolucion);
    }

}


// ─────────────────────────────────────────────
// ARMAR ESCENARIO
// Equivalente a armarEscenario() de Python.
// ─────────────────────────────────────────────

async function armarEscenario(canvas, lab, finalFil, finalCol) {
    const ctx = canvas.getContext('2d');

    dibujarLaberinto(lab);
    await sleep(1000);

    const posX   = BORDE + LADO / 2 + 1;
    const posY   = BORDE + LADO / 2 + 1;
    const posFil = 0;
    const posCol = 0;

    ctx.beginPath();
    ctx.arc(posX, posY, JUGADOR_RADIO, 0, Math.PI * 2);
    ctx.fillStyle = COLOR_JUGADOR;
    ctx.fill();

    const finalX = finalCol * LADO + JUGADOR_RADIO + BORDE + 1;
    const finalY = finalFil * LADO + JUGADOR_RADIO + BORDE + 1;
    ctx.fillStyle = COLOR_LLEGADA;
    ctx.fillRect(finalX, finalY, JUGADOR_DIAMETRO, JUGADOR_DIAMETRO);

    await sleep(1000);

    const distancia   = 0;
    const porAnalizar = new Map();
    porAnalizar.set(`${posFil},${posCol}`, distancia);
    const yaConocido = new Map();

    return { posX, posY, posFil, posCol, finalX, finalY, distancia, porAnalizar, yaConocido };
}


// ─────────────────────────────────────────────
// MOVER BUSCADOR
// Equivalente a mover() de Python.
// Tiene dos modos: exploración (COLOR_JUGADOR) y solución (COLOR_SOLUCION).
// ─────────────────────────────────────────────

async function moverBuscador(canvas, posX, posY, posFil, posCol, color) {
    const ctx  = canvas.getContext('2d');
    const antX = posX;
    const antY = posY;

    if (color === COLOR_JUGADOR) {
        // Efecto de huella semitransparente sobre el paso anterior
        // Equivalente al flashSurf con ALPHA_FONDO de Python
        ctx.fillStyle = ALPHA_FONDO;
        ctx.fillRect(posX - JUGADOR_RADIO, posY - JUGADOR_RADIO, JUGADOR_DIAMETRO, JUGADOR_DIAMETRO);
    }

    const nX = posCol * LADO + LADO / 2 + BORDE + 1;
    const nY = posFil * LADO + LADO / 2 + BORDE + 1;

    ctx.beginPath();
    ctx.arc(nX, nY, JUGADOR_RADIO, 0, Math.PI * 2);
    ctx.fillStyle = color;
    ctx.fill();

    if (color !== COLOR_JUGADOR) {
        // Línea conectora entre pasos de la solución
        ctx.strokeStyle = color;
        ctx.lineWidth   = DOBLEBORDE;
        ctx.lineCap     = 'round';
        ctx.beginPath();
        ctx.moveTo(antX, antY);
        ctx.lineTo(nX, nY);
        ctx.stroke();
    }

    await sleep(300);
    return { posX: nX, posY: nY };
}


// ─────────────────────────────────────────────
// PROCESAR RECORRIDO
// Equivalente a procesarRecorrido() de Python.
// Reconstruye el camino óptimo desde el Map yaConocido.
// En Python usaba popitem() que saca el último elemento del dict.
// En JS con Map hacemos lo mismo: sacamos la última entrada.
// ─────────────────────────────────────────────

function procesarRecorrido(yaConocido, lab) {
    // Sacar el último elemento (equivalente a popitem() de Python)
    const keys     = [...yaConocido.keys()];
    const lastKey  = keys[keys.length - 1];
    const distancia = yaConocido.get(lastKey);
    yaConocido.delete(lastKey);

    const [iFin, jFin] = lastKey.split(',').map(Number);
    const camino = [[iFin, jFin]];
    let nPasos   = 0;

    while (camino.length < distancia) {
        const entries = [...yaConocido.entries()];
        const lastEntry = entries[entries.length - 1];
        const [clave, _] = lastEntry;
        yaConocido.delete(clave);

        const [iInst, jInst] = clave.split(',').map(Number);
        const [iActual, jActual] = camino[nPasos];

        if (sonVecinos(iActual, jActual, iInst, jInst, lab)) {
            camino.push([iInst, jInst]);
            nPasos++;
            // Eliminar todos los items con la misma distancia
            const distInst = yaConocido.get(clave) ?? _;
            for (const [k, v] of [...yaConocido.entries()]) {
                if (v === distInst) yaConocido.delete(k);
            }
        }
    }

    camino.reverse();
    return camino;
}


// ─────────────────────────────────────────────
// SON VECINOS
// Equivalente a sonVecinos() de Python.
// ─────────────────────────────────────────────

function sonVecinos(fil1, col1, fil2, col2, lab) {
    if (fil1 - fil2 === 1 && col1 - col2 === 0) {
        if (!lab[fil1][col1].paredes.includes(ARRIBA) && !lab[fil2][col2].paredes.includes(ABAJO))
            return true;
    } else if (fil1 - fil2 === -1 && col1 - col2 === 0) {
        if (!lab[fil1][col1].paredes.includes(ABAJO) && !lab[fil2][col2].paredes.includes(ARRIBA))
            return true;
    } else if (fil1 - fil2 === 0 && col1 - col2 === 1) {
        if (!lab[fil1][col1].paredes.includes(IZQUIERDA) && !lab[fil2][col2].paredes.includes(DERECHA))
            return true;
    } else if (fil1 - fil2 === 0 && col1 - col2 === -1) {
        if (!lab[fil1][col1].paredes.includes(DERECHA) && !lab[fil2][col2].paredes.includes(IZQUIERDA))
            return true;
    }
    return false;
}


// ─────────────────────────────────────────────
// CAMINAR SOLUCIÓN
// Equivalente a caminarSolucion() de Python.
// ─────────────────────────────────────────────

async function caminarSolucion(canvas, camino) {
    const ctx  = canvas.getContext('2d');
    let posX = BORDE + LADO / 2 + 1;
    let posY = BORDE + LADO / 2 + 1;

    ctx.beginPath();
    ctx.arc(posX, posY, JUGADOR_RADIO, 0, Math.PI * 2);
    ctx.fillStyle = COLOR_SOLUCION;
    ctx.fill();

    for (const [fil, col] of camino) {
        ({ posX, posY } = await moverBuscador(canvas, posX, posY, fil, col, COLOR_SOLUCION));
    }

    await sleep(5000);
}


// ─────────────────────────────────────────────
// PRESENTAR RESULTADOS
// Equivalente a presentarResultados() de Python.
// Muestra una línea divisoria y las estadísticas de DFS vs BFS.
// ─────────────────────────────────────────────

async function presentarResultados(canvas, lab, DFStotal, DFSsolucion, BFStotal, BFSsolucion) {
    const ctx = canvas.getContext('2d');
    const msFrame = Math.floor(1000 / FPS);

    await sleep(1000);

    // Borrar paredes de a una (equivalente al loop borrarParedes de Python)
    for (let i = 0; i < N_FILAS; i++)
        for (let j = 0; j < N_COLUMNAS; j++)
            borrarParedes(i, j);

    ctx.fillStyle = COLOR_FONDO;
    ctx.fillRect(0, 0, ANCHO_PANTALLA, ALTO_PANTALLA);

    // Línea divisoria que crece desde arriba (igual que en Python)
    await new Promise(resolve => {
        let altoRect = 0;
        let dif = 1;

        function frame() {
            if (altoRect >= ALTO_PANTALLA) { resolve(); return; }

            altoRect += dif;
            dif += 2;
            altoRect = Math.min(altoRect, ALTO_PANTALLA);

            ctx.fillStyle = COLOR_PARED;
            ctx.fillRect(ANCHO_PANTALLA / 2, 0, BORDE, altoRect);

            requestAnimationFrame(frame);
        }
        requestAnimationFrame(frame);
    });

    await sleep(500);

    // Textos DFS (lado izquierdo)
    ctx.textAlign    = 'center';
    ctx.textBaseline = 'middle';

    ctx.font      = 'bold 32px monospace';
    ctx.fillStyle = COLOR_LLEGADA;
    ctx.fillText('DFS', ANCHO_PANTALLA / 4, ALTO_PANTALLA / 4);

    ctx.font = '24px monospace';
    ctx.fillText(`Pasos totales: ${DFStotal}`,              ANCHO_PANTALLA / 4, ALTO_PANTALLA / 2 - 30);
    ctx.fillText(`Pasos para llegar a destino: ${DFSsolucion}`, ANCHO_PANTALLA / 4, ALTO_PANTALLA / 2 + 40);

    // Textos BFS (lado derecho)
    ctx.font = 'bold 32px monospace';
    ctx.fillText('BFS', (ANCHO_PANTALLA / 4) * 3, ALTO_PANTALLA / 4);

    ctx.font = '24px monospace';
    ctx.fillText(`Pasos totales: ${BFStotal}`,              (ANCHO_PANTALLA / 4) * 3, ALTO_PANTALLA / 2 - 30);
    ctx.fillText(`Pasos para llegar a destino: ${BFSsolucion}`, (ANCHO_PANTALLA / 4) * 3, ALTO_PANTALLA / 2 + 40);

    await sleep(8000);
}

File: py/laberinto4_Inteligente.py
import random, pygame, sys
from pygame.locals import *
from laberinto0 import *


"""
LOS LABERINTOS COMO PUERTA DE ENTRADA A LA INTELIGENCIA ARTIFICAL

Aquí, el laberinto se resuelve solo (quiero decir, sin su ayuda). Aquí lo resuelven dos algoritmos: depth-first search (DFS) y breadth-first search (BFS).
El objetivo de este programa, es mostrar, de forma animada, cómo se comporta cada algoritmo en la misma situación, para poder compararlos.
Ambos algoritmos se ejecutan en los 5 tipos de laberintos existentes en laberinto0.py.

Algunas aclaraciones:

    1) Esto no es una implementación clásica. Es una implementación libre y adaptada a estos laberintos. Sin embargo, creo
        que resulta interesante para adentrarse en el tema.

    2) Una solución del laberinto se entiende como un camino que vaya desde el punto inicial hasta la meta (el cuadrado blanco).

    3) Para que se noten mejor las diferencias entre los dos algoritmos, en cada laberinto existen varias soluciones posibles.

"""


FPS = 24

BORDE = 4
DOBLEBORDE = BORDE * 2
ANCHO_LIENZO = 1000
ALTO_LIENZO = 600
ANCHO_PANTALLA = ANCHO_LIENZO + DOBLEBORDE
ALTO_PANTALLA = ALTO_LIENZO + DOBLEBORDE

assert ((BORDE % 2) == 0) and ((ANCHO_LIENZO % 2) == 0) and ((ALTO_LIENZO % 2) == 0), 'Convendría que las dimensiones sean números pares'

TAM_LADOS = [100, 100, 50, 50, 50]
CANT_NIVELES = len(TAM_LADOS)

for lado in TAM_LADOS:
    assert ((ALTO_LIENZO % lado) == 0) and ((ANCHO_LIENZO % lado) == 0), 'Convendría que las dimensiones del lienzo sean divisibles por el tamaño del lado del cuadrado'


COLOR_FONDO =   ( 20,  20,  20)
ALPHA_FONDO =   ( 20,  20,  20, 120)
COLOR_PARED =   ( 50, 200,  50)
COLOR_JUGADOR = (240, 200, 100)
COLOR_LLEGADA = (250, 250, 250)
COLOR_SOLUCION = (250, 160, 30)

ARRIBA = 'arriba'
ABAJO = 'abajo'
IZQUIERDA = 'izquierda'
DERECHA = 'derecha'


def main():
    global FPSCLOCK, DISPLAYSURF, LADO, N_FILAS, N_COLUMNAS, JUGADOR_RADIO, JUGADOR_DIAMETRO, FUENTE_TITULO, FUENTE_COMUN
    pygame.init()
    FPSCLOCK = pygame.time.Clock()
    DISPLAYSURF = pygame.display.set_mode((ANCHO_PANTALLA, ALTO_PANTALLA))
    pygame.display.set_caption('Laberinto')
    FUENTE_TITULO = pygame.font.Font('freesansbold.ttf', 32)
    FUENTE_COMUN = pygame.font.Font('freesansbold.ttf', 26)

    activarMultisalida()

    for nivel in range(len(TAM_LADOS)):

        chequearCierreDelPrograma()

        # Cada nivel o escenario tiene distinto tamaño, por eso se definen algunas dimensiones según el nivel.
        LADO, N_FILAS, N_COLUMNAS, JUGADOR_RADIO, JUGADOR_DIAMETRO = prepararEscenario((nivel), TAM_LADOS, ALTO_LIENZO, ANCHO_LIENZO)
        compartirGlobales(DISPLAYSURF, N_FILAS, N_COLUMNAS, BORDE, LADO, ANCHO_PANTALLA, ALTO_PANTALLA, COLOR_FONDO, COLOR_PARED, FPSCLOCK, FPS)
        # Asimismo, se genera un tipo de laberinto particular según el nivel.
        # Algo particular de este programa, es que genera laberintos más "abiertos" que los anteriores (ver la función generarLados()).
        # El objetivo de esto, es que haya más de una solución posible en cada laberinto, lo cual permite evidenciar un poco
        # las ventajas y las desventajas de cada algorimo (ya sea el DFS o el BFS)
        lab = generarLaberinto(nivel)
        # Se elige un punto final (más o menos) aleatorio, por la misma razón recién mencionada.
        finalFil, finalCol = random.randint(3, N_FILAS - 1), random.randint(3, N_COLUMNAS - 1)
        # Se inicializan los "contadores" o "estadísticas" de cada algoritmo
        DFStotal, DFSsolucion, BFStotal, BFSsolucion = 0, 0, 0, 0

        # La búsqueda se realiza dos veces: primero al estilo DFS (tipo_de_busqueda == 0), después al estilo BFS (tipo_de_busqueda == 1)
        for tipo_de_busqueda in range(2):
            chequearCierreDelPrograma()
            # Se inizializan alguas cosas básicas, como la posición del "jugador" o "buscador" y su "conocimiento inicial":
            # la distancia al lugar de origen es 0 (porque ya empieza en su lugar de origen),
            # el diccionario por_analizar simplemente contiene su lugar inicial
            # y el diccionario ya_conocido (encargado de registrar lo que ya se analizó) empieza vacío 
            posX, posY, posFil, posCol, finalX, finalY, distancia, por_analizar, ya_conocido = armarEscenario(lab, finalFil, finalCol)
            # Hasta no llegar a destino, se siguen buscando nuevos lugares para ser analizados.
            # En este contexto, ser analizado significa básicamente chequear si un luegar es o no es el destino final.
            while not (posFil == finalFil and posCol == finalCol):
                chequearCierreDelPrograma()
                del por_analizar[(posFil, posCol)]
                ya_conocido[(posFil, posCol)] = distancia

                if (posFil > 0):
                    if (not ARRIBA in lab[posFil][posCol][1]) and (not ABAJO in lab[posFil-1][posCol][1]) and (not (posFil - 1, posCol) in ya_conocido):
                        por_analizar[(posFil - 1, posCol)] = distancia + 1
                if (posFil < (N_FILAS - 1)):
                    if (not ABAJO in lab[posFil][posCol][1]) and (not ARRIBA in lab[posFil+1][posCol][1]) and (not (posFil + 1, posCol) in ya_conocido):
                        por_analizar[(posFil + 1, posCol)] = distancia + 1
                if (posCol > 0):
                    if (not IZQUIERDA in lab[posFil][posCol][1]) and (not DERECHA in lab[posFil][posCol-1][1]) and (not (posFil, posCol - 1) in ya_conocido):
                        por_analizar[(posFil, posCol - 1)] = distancia + 1
                if (posCol < (N_COLUMNAS - 1)):
                    if (not DERECHA in lab[posFil][posCol][1]) and (not IZQUIERDA in lab[posFil][posCol+1][1]) and (not (posFil, posCol + 1) in ya_conocido):
                        por_analizar[(posFil, posCol + 1)] = distancia + 1

                if tipo_de_busqueda == 0:
                    proximo = list(por_analizar.items())[-1] # Se toma el último que se agregó, tipo un stack. Esto hace que sea DFS
                else:
                    proximo = next(iter(por_analizar.items())) # Se toma el primero que hay, tipo una queue. Esto hace que sea BFS

                posFil, posCol, distancia = proximo[0][0], proximo[0][1], proximo[1]
                posX, posY = mover(posX, posY, posFil, posCol, COLOR_JUGADOR)

            ya_conocido[(posFil, posCol)] = distancia
            total = len(ya_conocido) - 1
            camino = procesarRecorrido(ya_conocido, lab)
            caminarSolucion(camino)

            if tipo_de_busqueda == 0:
                DFStotal, DFSsolucion = total, len(camino)
            else:
                BFStotal, BFSsolucion = total, len(camino)

        presentarResultados(lab, posX, posY, DFStotal, DFSsolucion, BFStotal, BFSsolucion)

    cerrarPrograma()




def presentarResultados(lab, posX, posY, DFStotal, DFSsolucion, BFStotal, BFSsolucion):
    pygame.time.wait(1000)
    pares_ordenados = []
    for i in range(N_FILAS):
        for j in range(N_COLUMNAS):
            borrarParedes(i, j)

    pygame.draw.rect(DISPLAYSURF, COLOR_FONDO, (0, 0, ANCHO_PANTALLA, ALTO_PANTALLA))

    alto_rect = 0
    dif = 1
    while (alto_rect < ALTO_PANTALLA):
        chequearCierreDelPrograma()
        alto_rect += dif
        dif += 2
        pygame.draw.rect(DISPLAYSURF, COLOR_PARED, (ANCHO_PANTALLA // 2, 0, BORDE, alto_rect))
        pygame.display.update()
        FPSCLOCK.tick(FPS)
    pygame.time.wait(500)

    textSurfaceObj = FUENTE_TITULO.render("DFS", True, COLOR_LLEGADA)
    textRectObj = textSurfaceObj.get_rect()
    textRectObj.center = (ANCHO_PANTALLA // 4, ALTO_PANTALLA // 4)
    DISPLAYSURF.blit(textSurfaceObj, textRectObj)

    textSurfaceObj = FUENTE_COMUN.render(f"Pasos totales: {DFStotal}", True, COLOR_LLEGADA)
    textRectObj = textSurfaceObj.get_rect()
    textRectObj.center = (ANCHO_PANTALLA // 4, ALTO_PANTALLA // 2 - 30)
    DISPLAYSURF.blit(textSurfaceObj, textRectObj)

    textSurfaceObj = FUENTE_COMUN.render(f"Pasos para llegar a destino: {DFSsolucion}", True, COLOR_LLEGADA)
    textRectObj = textSurfaceObj.get_rect()
    textRectObj.center = (ANCHO_PANTALLA // 4, ALTO_PANTALLA // 2 + 40)
    DISPLAYSURF.blit(textSurfaceObj, textRectObj)

    textSurfaceObj = FUENTE_TITULO.render("BFS", True, COLOR_LLEGADA)
    textRectObj = textSurfaceObj.get_rect()
    textRectObj.center = ((ANCHO_PANTALLA // 4) * 3, ALTO_PANTALLA // 4)
    DISPLAYSURF.blit(textSurfaceObj, textRectObj)

    textSurfaceObj = FUENTE_COMUN.render(f"Pasos totales: {BFStotal}", True, COLOR_LLEGADA)
    textRectObj = textSurfaceObj.get_rect()
    textRectObj.center = ((ANCHO_PANTALLA // 4) * 3, ALTO_PANTALLA // 2 - 30)
    DISPLAYSURF.blit(textSurfaceObj, textRectObj)

    textSurfaceObj = FUENTE_COMUN.render(f"Pasos para llegar a destino: {BFSsolucion}", True, COLOR_LLEGADA)
    textRectObj = textSurfaceObj.get_rect()
    textRectObj.center = ((ANCHO_PANTALLA // 4) * 3, ALTO_PANTALLA // 2 + 40)
    DISPLAYSURF.blit(textSurfaceObj, textRectObj)

    pygame.display.update()
    FPSCLOCK.tick(FPS)
    for _ in range(7):
        chequearCierreDelPrograma()
        pygame.time.wait(1000)


def armarEscenario(lab, finalFil, finalCol):
    # Se dibuja el laberinto
    dibujarLaberinto(lab)
    pygame.time.wait(1000)
    # Se determina la posición inicial del jugador, en pixeles y en la matriz
    posX, posY, posFil, posCol = BORDE + LADO / 2 + 1, BORDE + LADO / 2 + 1, 0, 0
    pygame.draw.circle(DISPLAYSURF, COLOR_JUGADOR, (posX, posY), JUGADOR_RADIO)
    # Se dibuja el punto de llegada
    finalX, finalY = (finalCol * LADO) + JUGADOR_RADIO + BORDE + 1, (finalFil * LADO) + JUGADOR_RADIO + BORDE + 1
    pygame.draw.rect(DISPLAYSURF, COLOR_LLEGADA, (finalX, finalY, JUGADOR_DIAMETRO, JUGADOR_DIAMETRO))
    # Se actualiza la pantalla
    pygame.display.update()
    pygame.time.wait(1000)
    # Se inicializan los elementos necesarios para llevar adelante la búsqueda
    distancia = 0
    por_analizar = {(posFil, posCol): distancia}
    ya_conocido = {}
    return posX, posY, posFil, posCol, finalX, finalY, distancia, por_analizar, ya_conocido


# Hay dos tipos de movimientos: cuando se está explorando (en cuyo caso, el color es COLOR_JUGADOR) y cuando ya se encontró la solución.
# Cuando se está explorando, no se borra el paso anterior, pero sí se lo tapa un poco, como si quedara una huella.
# Cuando se recorre la solución, se deja cada paso bien pintado de naranja y se agrega un conector (una línea) entre los dos pasos.
def mover(posX, posY, posFil, posCol, color):
    antX, antY = posX, posY
    if color == COLOR_JUGADOR:
        rectangle = pygame.Rect(posX - JUGADOR_RADIO, posY - JUGADOR_RADIO, JUGADOR_DIAMETRO, JUGADOR_DIAMETRO)
        origSurf = DISPLAYSURF.copy()
        flashSurf = pygame.Surface((JUGADOR_DIAMETRO, JUGADOR_DIAMETRO))
        flashSurf = flashSurf.convert_alpha()
        DISPLAYSURF.blit(origSurf, (0, 0))
        flashSurf.fill(ALPHA_FONDO)
        DISPLAYSURF.blit(flashSurf, rectangle.topleft)

    posX, posY = (posCol * LADO) + (LADO / 2) + BORDE + 1, (posFil * LADO) + (LADO / 2) + BORDE + 1
    pygame.draw.circle(DISPLAYSURF, color, (posX, posY), JUGADOR_RADIO)

    if color != COLOR_JUGADOR:
        pygame.draw.line(DISPLAYSURF, color, (antX, antY), (posX, posY), DOBLEBORDE)

    pygame.display.update()
    pygame.time.wait(300)
    return posX, posY


# Esta función recibe un diccionario ("ya_conocido") donde cada clave es una posición única (fil, col) y cada valor es algún entero
# que representa la distancia o el número de pasos necesario para llegar hasta ahí (según el camino que se tomó, no necesariamente es la distancia mínima).
# Lo que hace la función es buscar los pasos, desde el final hasta el principio, guardarlos en una lista,
# y luego invertir esa lista para que queden en orden.
# Para buscar los pasos desde el final hasta el principio, busca un cubículo vecino y, una vez que lo encuentra,
# elimina del diccionario aquellos items que tengan esa misma distancia. Repite este proceso hasta que
# la cantidad de pasos sea igual a la distancia a la que se encuentra el objetivo con respecto al punto inicial, o sea, hasta que len(camino) == distancia.
# Esta parte del programa es quizá la única parte que se diferencia de las implementaciones tradicionales de DFS y BFS.
# Pero bueno, lo importante es que esta función recibe el diccionario "ya_conocido" y devuelve la solución del laberinto.
def procesarRecorrido(ya_conocido, lab):
    instancia = ya_conocido.popitem()
    distancia = instancia[1]
    camino = [instancia[0]]
    n_pasos = 0
    while len(camino) < distancia:
        instancia = ya_conocido.popitem()
        if sonVecinos(camino[n_pasos][0], camino[n_pasos][1], instancia[0][0], instancia[0][1], lab):
            camino.append(instancia[0])
            n_pasos += 1
            ya_conocido = {key: value for key, value in ya_conocido.items() if value != instancia[1]}

    camino.reverse()
    return camino


# Para que un paso entre dos puntos sea "legítimo", deben cumplirse dos condiciones:
#   1) los puntos deben estar o bien a una fila o bien a una columna de distancia;
#   2) no debe haber ninguna pared entre ellos.
def sonVecinos(fil1, col1, fil2, col2, lab):
    if (fil1 - fil2 == 1) and (col1 - col2 == 0):
        if (ARRIBA not in lab[fil1][col1][1]) and (ABAJO not in lab[fil2][col2][1]):
            return True
    elif (fil1 - fil2 == -1) and (col1 - col2 == 0):
        if (ABAJO not in lab[fil1][col1][1]) and (ARRIBA not in lab[fil2][col2][1]):
            return True
    elif (fil1 - fil2 == 0) and (col1 - col2 == 1):
        if (IZQUIERDA not in lab[fil1][col1][1]) and (DERECHA not in lab[fil2][col2][1]):
            return True
    elif (fil1 - fil2 == 0) and (col1 - col2 == -1):
        if (DERECHA not in lab[fil1][col1][1]) and (IZQUIERDA not in lab[fil2][col2][1]):
            return True
    else:
        return False


def caminarSolucion(camino):
    posX, posY, posFil, posCol = BORDE + LADO / 2 + 1, BORDE + LADO / 2 + 1, 0, 0
    pygame.draw.circle(DISPLAYSURF, COLOR_SOLUCION, (posX, posY), JUGADOR_RADIO)
    pygame.display.update()
    for paso in camino:
        chequearCierreDelPrograma()
        posX, posY = mover(posX, posY, paso[0], paso[1], COLOR_SOLUCION)
    for _ in range(5):
        chequearCierreDelPrograma()
        pygame.time.wait(1000)


if __name__ == '__main__':
    main()

